crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,320 @@
+        aws_config_builder = aws_config_builder.profile_name(profile);
+    };
+    let sdk_config = aws_config_builder.load().await;
+
+    let mut builder = aws_sdk_s3::config::Builder::from(&sdk_config);
+    if let Some(force_path_style) = force_path_style {
+        builder = builder.force_path_style(force_path_style);
+    };
+    let s3_config = builder.build();
+
+    let client: aws_sdk_s3::Client = aws_sdk_s3::Client::from_conf(s3_config);
+    client
+}
+
+impl S3Middleware {
+    /// Create a new S3 middleware.
+    pub fn new(
+        config_file: Option<PathBuf>,
+        profile: Option<String>,
+        force_path_style: Option<bool>,
+    ) -> Self {
+        Self {
+            config_file,
+            profile,
+            force_path_style,
+            expiration: std::time::Duration::from_secs(300),
+        }
+    }
+
+    /// Generate a presigned S3 `GetObject` request.
+    async fn generate_presigned_s3_request(
+        &self,
+        bucket_name: &str,
+        key: &str,
+    ) -> MiddlewareResult<PresignedRequest> {
+        let client = create_client(
+            self.config_file.clone(),
+            self.profile.clone(),
+            self.force_path_style,
+        )
+        .await;
+        let builder = client.get_object().bucket(bucket_name).key(key);
+        builder
+            .presigned(
+                PresigningConfig::expires_in(self.expiration)
+                    .map_err(reqwest_middleware::Error::middleware)?,
+            )
+            .await
+            .map_err(reqwest_middleware::Error::middleware)
+    }
+}
+
+#[async_trait]
+impl Middleware for S3Middleware {
+    /// Create a new authentication middleware for S3.
+    async fn handle(
+        &self,
+        mut req: Request,
+        extensions: &mut http::Extensions,
+        next: Next<'_>,
+    ) -> MiddlewareResult<Response> {
+        if req.url().scheme() == "s3" {
+            let url = req.url().clone();
+            let bucket_name = url.host_str().expect("Host should be present in S3 URL");
+            let key = url.path();
+            let presigned_request = self.generate_presigned_s3_request(bucket_name, key).await?;
+
+            *req.url_mut() = Url::parse(presigned_request.uri())
+                .map_err(reqwest_middleware::Error::middleware)?;
+        }
+        next.run(req, extensions).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use super::*;
+    use rstest::{fixture, rstest};
+    use serial_test::serial;
+    use tempfile::{tempdir, TempDir};
+
+    async fn with_env(
+        env: HashMap<&str, &str>,
+        f: impl FnOnce() -> std::pin::Pin<Box<dyn std::future::Future<Output = ()> + Send>>,
+    ) {
+        for (key, value) in &env {
+            std::env::set_var(key, value);
+        }
+        f().await;
+        for (key, _) in env {
+            std::env::remove_var(key);
+        }
+    }
+
+    #[tokio::test]
+    #[serial]
```

Comment:
these tests need to be serial in order for the env variables not overwriting each other

Reply:
Lets use this crate instead: https://crates.io/crates/temp-env